src/erc721/ERC721.sol

```diff
@@ -0,0 +1,193 @@
+
+    mapping(address => uint256) public nonces;
+    
+    constructor(
+        string memory _name,
+        string memory _symbol
+    ) {
+        name = _name;
+        symbol = _symbol;
+
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
+                keccak256(bytes(name)),
+                keccak256(bytes("1")),
+                block.chainid,
+                address(this)
+            )
+        );
+    }
+    
+    /*///////////////////////////////////////////////////////////////
+                              ERC-721 LOGIC
+    //////////////////////////////////////////////////////////////*/
+    
+    function supportsInterface(bytes4 sig) external pure returns (bool) {
+        return (sig == 0x80ac58cd || sig == 0x5b5e139f);
+    }
+    
+    function approve(address spender, uint256 tokenId) external {
+        address owner = ownerOf[tokenId];
+        
+        require(msg.sender == owner || isApprovedForAll[owner][msg.sender], "NOT_AUTHORIZED");
+        
+        getApproved[tokenId] = spender;
+        
+        emit Approval(msg.sender, spender, tokenId); 
+    }
+    
+    function setApprovalForAll(address operator, bool approved) external {
+        isApprovedForAll[msg.sender][operator] = approved;
+        
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function transfer(address to, uint256 tokenId) external {
+        require(msg.sender == ownerOf[tokenId], "NOT_OWNER");
+        
+        balanceOf[msg.sender]--; 
+        
+        balanceOf[to]++; 
+        
+        getApproved[tokenId] = address(0);
+        
+        ownerOf[tokenId] = to;
+        
+        emit Transfer(msg.sender, to, tokenId); 
+    }
+    
+    function transferFrom(address, address to, uint256 tokenId) external {
+        address owner = ownerOf[tokenId];
+        
+        require(
+            msg.sender == owner || getApproved[tokenId] == msg.sender 
+            || isApprovedForAll[owner][msg.sender], "NOT_AUTHORIZED"
+        );
+        
+        balanceOf[owner]--; 
+        
+        balanceOf[to]++;
+        
+        getApproved[tokenId] = address(0);
+        
+        ownerOf[tokenId] = to;
+        
+        emit Transfer(owner, to, tokenId); 
+    }
+    
+    /*///////////////////////////////////////////////////////////////
+                          PERMIT/EIP-2612 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function permit(
+        address owner,
+        address spender,
+        uint256 tokenId,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        require(deadline >= block.timestamp, "PERMIT_DEADLINE_EXPIRED");
+        // This is reasonably safe because incrementing past type(uint256).max
+        // through approvals is exceedingly unlikely!
+        unchecked {
```

Comment:
can we limit the unchecked to just where the nonce is incremented (idk if u can wrap just around nonces++ but at least only the digest block

Reply:
also should do this with ERC20 if we're doing here

Reply:
mehhh idk i'd rather just remove this unchecked. not worth it here imo— all the other instances we use it at are sound

Reply:
yeah, so the `unchecked` bracketing seemingly scopes away the `digest` variable, so need to wrap underlying. But agree, might as well keep things totally safe -- will remove :~)